### PR TITLE
WIP: RFC: pkg/assets: Merkle DAG asset library

### PR DESCRIPTION
**Pull request by wking**
Our current `pkg/asset` approach requires multiple stages to run, and selecting those stages requires some dev planning and user education.  This new package allows us to dump the whole asset graph at once, and then edit/rebuild multiple times while maintaining a consistent asset state.  I *think* this will be easier for folks to wrap their heads around, but I'm not always the best judge of approachability ;).  Check out the updated user docs (included in the PR) for an overview of the new approach.

If folks think this approach is worthwhile, I can keep working through the process of porting our old assets over to the new framework.  Currently there are only a handful as a proof-of-concept.

### CODE REVIEW

**Comment by abhinavdahiya** [comment]
> There's a dependency graph. And besides the mostly-complete one you'll be able to find in the docs, you can now run graph against an asset directory to see your dependencies.

:/

> The files don't need to go anywhere between clusters. Once you're happy with create assets, you'll be able to cycle between create cluster and destroy cluster without needing to return to create assets. I think this will reduce the chances of users doing things like` removing their state while leaving a branch asset like their kubeconf around (#522, #532).

so this requires users keep the whole dir around for that, i don't see how that is differnet that state file.
https://github.com/openshift/installer/pull/547 shows current approach can be successfull in same action.

> I think there's simplicity in dropping the target and writeable-asset distinctions. Now there is only one rendering command and only one type of asset.

I personally have been trying my best to keep us from adding new interfaces too, but targets are a better UX

> This allows me to colocate the rebuilder/defaulter definitions with their registrations. If you'd rather have me centralize the registrations in one big, non-init var, I could do that. But note that ConstantDefault registries for some scripts and templates will be spacious.

The current impl is able to work without any init

> It's all bytes on disk for both this and our current approach. If you want helpers to unpack a particular asset into a struct for its children

Current impl is slowly moving away from bytes... https://github.com/openshift/installer/blob/master/pkg/asset/machines/worker.go#L69 is accessing the installconfig type directly, not marshal/unmarshalling, while this forces us to go back

> That's a bonus, right? Why would you want custom read logic? ;)

reading expired certs from disk etc validations can be useful to user.

> We ignore future changes to its ancestors, we continue to propagate down to its decendants. If we wanted, we could record its parent references when we freeze it and log warnings if those parents changed (e.g. "you've added a custom kubeconfig, but its previous parent root-ca.crt has changed from old-hash to new-hash", etc.).

The current impl can handle that case easily too, disregard parents when asset from disk is read and modified...

Also how do we model a case where user wants to bring new files to an asset, eg. manfiests from current impl.

**Comment by wking** [comment]
I've pushed dd9b8cc -> 20912ea, rebasing onto master and adding enough assets to generate a reasonable [graph][1].

> > The files don't need to go anywhere between clusters. Once you're happy with create assets, you'll be able to cycle between create cluster and destroy cluster without needing to return to create assets. I think this will reduce the chances of users doing things like` removing their state while leaving a branch asset like their kubeconf around (#522, #532).
>
> so this requires users keep the whole dir around for that, i don't see how that is differnet that state file.
> #547 shows current approach can be successfull in same action.

Yeah, #547 will give us similar behaviour here.  The difference is that with #547 you still need to find the right target to access the resource(s) you want to manipulate, while with this approach you (re)generate all the assets at once.

> > I think there's simplicity in dropping the target and writeable-asset distinctions. Now there is only one rendering command and only one type of asset.
>
> I personally have been trying my best to keep us from adding new interfaces too, but targets are a better UX

Because...?  It it just that there's less to think about at each stage?  One extension I was thinking about was:

```
openshift-install [options] create assets [TARGET ...]
```

to allow you to create subsets of the graph.  For example, you could:

```sh
openshift-install --dir example create assets bootstrap.ign
```

to just create assets through to the `bootstrap.ign` target.  A subsequent `... create assets` call would fill in the rest of the graph.  This would allow users who wanted separate stages to recover that behavior for a single intermediate step.  There wouldn't be any of the current master's removing of "consumed" asset files, so subsequent explicit targets would still have asset files from the previous iteration.

> > This allows me to colocate the rebuilder/defaulter definitions with their registrations. If you'd rather have me centralize the registrations in one big, non-init var, I could do that. But note that ConstantDefault registries for some scripts and templates will be spacious.
>
> The current impl is able to work without any init

Because the current implementation requires consumers to explicitly import their parents (e.g. [here][2]).  With this PR, you just feed your parents' names into `getByName` (e.g. [here][3]).  Not requiring a separate dependency lister allows for dynamic dependencies (e.g. you can pull the `platform` asset, see that it is `libvirt` and skip your potential AWS and OpenShift parents).  But you could still get dynamic dependencies by collecting all the registration in a single function, or in per-package functions (e.g. `tlsDefaults() map[string]Defaulter` and `tlsRebuilders() map[string]assets.Rebuild` if your issue is just with the use of `init`.

> > It's all bytes on disk for both this and our current approach. If you want helpers to unpack a particular asset into a struct for its children
>
> Current impl is slowly moving away from bytes... https://github.com/openshift/installer/blob/master/pkg/asset/machines/worker.go#L69 is accessing the installconfig type directly, not marshal/unmarshalling, while this forces us to go back

It's still unmarshalling when we pull it up off the disk.  The savings are just efficiency for assets that are consumed in multiple places.  I don't think we have so many of those that unmarshalling efficiency is going to be a big deal, but I can run some benchmarks if you want once I get the asset tree filled in a bit more.

> >    That's a bonus, right? Why would you want custom read logic? ;)
>
> reading expired certs from disk etc validations can be useful to user.

Validation sounds useful.  Adding a new `Asset.Validate` property could address that, no?

> > We ignore future changes to its ancestors, we continue to propagate down to its decendants. If we wanted, we could record its parent references when we freeze it and log warnings if those parents changed (e.g. "you've added a custom kubeconfig, but its previous parent root-ca.crt has changed from old-hash to new-hash", etc.).
>
> The current impl can handle that case easily too, disregard parents when asset from disk is read and modified...

Right, currently it's [disregarding the from-disk asset when a parent changes][4], but we could change that.  Looking through master's `store.go` though, I don't see any dirty flag being persisted in the state (like I'm [persisting][5] [`frozen`][6] ).  Am I missing something?  What happens if you want to drop in a custom worker ignition (say) and have it persist across multiple cluster launches?  Or, with the current coarse-grained assets in master, is there a way to persist your custom `kube-apiserver-config-overrides.yaml` across multiple cluster launches?

The current asset approach requires a lot of Go overhead for each asset, while the approach I have here requires very little.  For example, compare [this 29-line file for the kube CA][7] with [the current 44 liner][8].  And I have the bootstrap units a [separate assets][9] injected with [one line each][10], while the current master [pushes them into the bootstrap asset][11], presumthoably because the overhead of supporting [`Dependencies`, `Generate`, `Name`, `Files`, and `Load`][12] for each asset is too much trouble. 

> Also how do we model a case where user wants to bring new files to an asset, eg. manfiests from current impl.

Yeah, there's not an elegant way to do that.  We could hack something in for manifests in particular (do we need this for other asset types beyond manifests?).  As this branch stands, the easiest approach is probably stuffing your manifest into `bootstrap.ign.data` via `jq` or similar after each asset rebuild (which isn't bad).  But you'd need to remember to remove `bootstrap.ign` before each rebuild to clear your freeze and pick up parent changes.

[1]: https://github.com/openshift/installer/blob/20912eaaf4cd8f9311c73ec918d808689247bcfd/docs/user/assets.svg
[2]: https://github.com/openshift/installer/blob/762ae17f1176c5b835a3343218405bddd6123d5a/pkg/asset/ignition/bootstrap/bootstrap.go#L54-L74
[3]: https://github.com/openshift/installer/blob/20912eaaf4cd8f9311c73ec918d808689247bcfd/pkg/installerassets/ignition.go#L55-L67
[4]: https://github.com/openshift/installer/blob/762ae17f1176c5b835a3343218405bddd6123d5a/pkg/asset/store.go#L181-L182
[5]: https://github.com/openshift/installer/blob/20912eaaf4cd8f9311c73ec918d808689247bcfd/pkg/assets/asset.go#L163-L165
[6]: https://github.com/openshift/installer/blob/20912eaaf4cd8f9311c73ec918d808689247bcfd/pkg/assets/asset.go#L46-L48
[7]: https://github.com/openshift/installer/blob/20912eaaf4cd8f9311c73ec918d808689247bcfd/pkg/installerassets/tls/kubeca.go
[8]: https://github.com/openshift/installer/blob/762ae17f1176c5b835a3343218405bddd6123d5a/pkg/asset/tls/kubeca.go
[9]: https://github.com/openshift/installer/blob/20912eaaf4cd8f9311c73ec918d808689247bcfd/pkg/installerassets/systemd.go
[10]: https://github.com/openshift/installer/blob/20912eaaf4cd8f9311c73ec918d808689247bcfd/pkg/installerassets/ignition.go#L59-L62
[11]: https://github.com/openshift/installer/blob/762ae17f1176c5b835a3343218405bddd6123d5a/pkg/asset/ignition/bootstrap/bootstrap.go#L102-L105
[12]: https://github.com/openshift/installer/blob/762ae17f1176c5b835a3343218405bddd6123d5a/pkg/asset/asset.go#L11-L34

**Comment by wking** [comment]
I've pushed 0222815 -> 49efd2d ([new graph][1]) implementing [asset subdirectories][2].  The generated asset state (still very incomplete) now looks like:

```console
$ openshift-install --dir=wking create assets --prune
$ tree wking
wking
├── auth
│   ├── kubeconfig-admin
│   └── kubeconfig-kubelet
├── base-domain
├── bin
│   ├── bootkube.sh
│   └── report-progress.sh
├── cluster
├── cluster-name
├── ignition
│   └── bootstrap.ign
├── ignition-configs
├── manifests
│   ├── cluster-apiserver-certs.yaml
│   └── kube-apiserver-secret.yaml
├── overrides
│   ├── kube-apiserver-config-overrides.yaml
│   └── kube-controller-manager-config-overrides.yaml
├── ssh.pub
├── tls
│   ├── admin-client.crt
│   ├── admin-client.key
│   ├── aggregator-ca.crt
│   ├── aggregator-ca.key
│   ├── cluster-apiserver-ca.crt
│   ├── cluster-apiserver-ca.key
│   ├── kube-ca.crt
│   ├── kube-ca.key
│   ├── kubelet-client.crt
│   ├── kubelet-client.key
│   ├── root-ca.crt
│   └── root-ca.key
└── unit
    ├── bootkube.service
    ├── kubelet.service
    ├── progress.service
    └── tectonic.service

7 directories, 30 files
```

Folks interested in install-config level settings can just look at:

```console
$ ls -F --group-directories-first wking | cat
auth/
bin/
ignition/
manifests/
overrides/
tls/
unit/
base-domain
cluster
cluster-name
ignition-configs
ssh.pub
```

[1]: https://github.com/openshift/installer/blob/49efd2d84a7005431b4bf1e72d454dd8432bac7c/docs/user/assets.svg
[2]: https://github.com/openshift/installer/pull/556#issuecomment-433677710

**Comment by abhinavdahiya** [comment]
https://github.com/openshift/installer/pull/556/commits/58b31c2b73920d7a085206f294e8a7972decfeb9 I don't think this is the best way to allow any kind of review. i can't compare or verify if your copied asset impl is correct. please make changes inplace.

File: pkg/assets/asset.go [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+			return nil, err
+		}
+		parents[name] = &prnt
+
+		hash, err := parents[name].Hash()
+		if err != nil {
+			return nil, errors.Wrapf(err, "hash %q parent %q", asset.Name, name)
+		}
+
+		asset.Parents = append(asset.Parents, Reference{
```
[the remark this answers is not in the document]
Comment (wking) [reply pkg/assets/asset.go]: > this is open to duplications?

It is.  I'm comfortable avoiding that by not requesting the same parent twice from a single rebuilder.  But if you'd prefer a guard to catch folks who do so by mistake, or if you'd like me to *allow* repeat lookups without duplicate additions, I can put either of those in.

**Comment by wking** [comment]
> 58b31c2 I don't think this is the best way to allow any kind of review. i can't compare or verify if your copied asset impl is correct. please make changes inplace.

I could squash things down, but to get an in-place pivot, I'd need to squash all of:

```
* 58b31c2 pkg/asset: Remove now that we have pkg/installerassets
* e2ca6ea cmd/openshift: Transition to the new asset graph
* f646d01 pkg/installerassets: Use the generic asset package
```

together (and possibly also the doc updates currently in 813ea6e).  I can do that, you can see pretty much the same diff via [GitHub's file view][1], and I think that in general I'm mutating files too much for Git to be able to see things like the `pkg/asset/tls/root.go` -> `pkg/installerassets/tls/root.go`.

Or are you saying I should rename my files to preserve the old filenames where possible?  I could do that, but I'd like to keep the generic `pkg/assets` (59b9fa6) separate from our installer-specific assets (which I'm currently adding in `pkg/installerassets`).  And having both `package/assets` and `package/asset` side by side seems confusing ;).  Maybe you want `package/assetframework` or some such with the generic stuff, and `pkg/asset` for the installer-specific stuff?

> also see the dev docs on asset generation were dropped, i think they were useful as they allow any body new to look and add new asset.

The user-level portion of this doc moved into [`docs/user/overview.md`][2].  I thought `pkg/installerassets` gave enough examples of injecting assets for it to not need additional docs, but I can add them if you want (by fleshing out [these][3] with some examples and associated patter?

> Also there is no mention of `Defaults` vs `Rebuilder` that you have registered, why, when, or to use them?

Because users don't need to care about them, and devs can read the (currently sparse) godocs.  I'll take a stab at extending the godocs, and we'll see how that feels to you.

[1]: https://github.com/openshift/installer/pull/556/files/33e2c008a3452eb8396bdf98ebdef47b61c011d0..58b31c2b73920d7a085206f294e8a7972decfeb9
[2]: https://github.com/openshift/installer/blob/813ea6e67dca9c26acb73a2ad89c14efb5e57734/docs/user/overview.md#asset-generation
[3]: https://github.com/openshift/installer/blob/f646d01608bfe742d25760e5e79dec13a0b71ffa/pkg/installerassets/installerassets.go#L12-L19

**Comment by wking** [comment]
I've rebased onto master with 49efd2d -> 5994f8b ([new graph][1]).  The graph nodes are now colored by directory to make it easier to get a quick overview.  I've also [added more docs][2] to help orient devs:

```console
$ godoc ./pkg/installerassets
...
VARIABLES

var Defaults = make(map[string]Defaulter)
    Defaults registers installer asset default functions by name. Use this
    to set up assets that do not have parents. For example, constants:

	Defaults["your/asset"] = ConstantDefault([]byte("your value"))

    or values populated from outside the asset graph:

	Defaults["your/asset"] = func() (data []byte, err error) {
	  value = os.Getenv("YOUR_ENVIRONMENT_VARIABLE")
	  return []byte(value), nil
	}

var Rebuilders = make(map[string]assets.Rebuild)
    Rebuilders registers installer asset rebuilders by name. Use this to set
    up assets that have parents. For example:

	func yourRebuilder(getByName assets.GetByString) (asset *assets.Asset, err error) {
	  asset = &assets.Asset{
	    Name:          "tls/root-ca.crt",
	    RebuildHelper: rootCARebuilder,
	  }

	  parents, err := asset.GetParents(getByName, "tls/root-ca.key")
	  if err != nil {
	    return nil, err
	  }

	  // Assemble your data based on the parent content using your custom logic.
	  for name, parent := range parents {
	    asset.Data = append(asset.Data, parent.Data)
	  }

	  return asset, nil
	}

    and then somewhere (e.g. an init() function), add it to the registry:

	Rebuilders["your/asset"] = yourRebuilder
...
```

[1]: https://github.com/openshift/installer/blob/5994f8b6976e3ed63a0749a2ae057dfd2e5ba9ab/docs/user/assets.svg
[2]: https://github.com/openshift/installer/pull/556#issuecomment-434048915

**Comment by wking** [comment]
I've pushed 5994f8b -> d3685ba adding a few more assets and minor fixes.  There's also a user-facing change to get log entries about freezing parents:

```console
$ go test -v ./pkg/assets
...
    	asset.go:191: "b" data changed via "/tmp/openshift-install-037749922/b", future changes to the previous parents (a/a) will no longer update this asset
...
```

We could push a stronger error if we recorded `Asset.FreezeTimeParents` and warned if any of them had has mismatches, but I haven't gone that far yet.

**Comment by abhinavdahiya** [comment]
> I've rebased onto master with [49efd2d](https://github.com/openshift/installer/commit/49efd2d84a7005431b4bf1e72d454dd8432bac7c) -> [5994f8b](https://github.com/openshift/installer/commit/5994f8b6976e3ed63a0749a2ae057dfd2e5ba9ab) ([new graph](https://github.com/openshift/installer/blob/5994f8b6976e3ed63a0749a2ae057dfd2e5ba9ab/docs/user/assets.svg)). The graph nodes are now colored by directory to make it easier to get a quick overview. I've also [added more docs](https://github.com/openshift/installer/pull/556#issuecomment-434048915) to help orient devs:
> 
> ```
> $ godoc ./pkg/installerassets
> ...
> VARIABLES
> 
> var Defaults = make(map[string]Defaulter)
>     Defaults registers installer asset default functions by name. Use this
>     to set up assets that do not have parents. For example, constants:
> 
> 	Defaults["your/asset"] = ConstantDefault([]byte("your value"))
> 
>     or values populated from outside the asset graph:
> 
> 	Defaults["your/asset"] = func() (data []byte, err error) {
> 	  value = os.Getenv("YOUR_ENVIRONMENT_VARIABLE")
> 	  return []byte(value), nil
> 	}
> 
> var Rebuilders = make(map[string]assets.Rebuild)
>     Rebuilders registers installer asset rebuilders by name. Use this to set
>     up assets that have parents. For example:
> 
> 	func yourRebuilder(getByName assets.GetByString) (asset *assets.Asset, err error) {
> 	  asset = &assets.Asset{
> 	    Name:          "tls/root-ca.crt",
> 	    RebuildHelper: rootCARebuilder,
> 	  }
> 
> 	  parents, err := asset.GetParents(getByName, "tls/root-ca.key")
> 	  if err != nil {
> 	    return nil, err
> 	  }
> 
> 	  // Assemble your data based on the parent content using your custom logic.
> 	  for name, parent := range parents {
> 	    asset.Data = append(asset.Data, parent.Data)
> 	  }
> 
> 	  return asset, nil
> 	}
> 
>     and then somewhere (e.g. an init() function), add it to the registry:
> 
> 	Rebuilders["your/asset"] = yourRebuilder
> ...
> ```

This seems like we can collapse this distinction, node with no parent are still nodes in graph? This is un-necessary optimization... ?

**Comment by wking** [comment]
> > ```
> > var Rebuilders = make(map[string]assets.Rebuild)
> >     Rebuilders registers installer asset rebuilders by name. Use this to set
> >     up assets that have parents. For example:
> > ```
>
> This seems like we can collapse this distinction, node with no parent are still nodes in graph? This is un-necessary optimization... ?

It's seemed convenient from a dev perspective to go this way, but yeah, we could drop `Defaults` and just have `Rebuilders` with a:

```
func NoParentRebuilder(name string, defaulter Defaulter) assets.Rebuild {
  return func(asset *assets.Asset, err error) {
    return &assets.Asset{
      Name: name,
      Data: defaulter(),
      RebuildHelper: NoParentRebuilder(
    }, nil
  }
}
```

with:

```go
installerassets.Rebuilders["tls/root-ca.key"] = NoParentRebuilder("tls/root-ca.key", privateKey)
```

instead of my [current][1]:

```go
installerassets.Defaults["tls/root-ca.key"] = privateKey
```

I felt like this case was common enough that adding `Defaults` was worth it, but I can drop it if you like.

[1]: https://github.com/openshift/installer/commit/5652fecaf732d82a5bc79a14345b747135c76a92#diff-e13c68e2caeebe4a1c0cfb2f0bfa0cc4R61

**Comment by abhinavdahiya** [comment]
> I felt like this case was common enough that adding Defaults was worth it, but I can drop it if you like.

the less we differ and specialize the better. 
what would happen regarding these ending up on disk?

**Comment by wking** [comment]
I've rebased on master and [squashed the three pivot commits together][1] with d3685ba -> 0fb0b46.  There are fewer in-Go constants now that I'm building on top of #510 and #592.  In 0fb0b46, I've shifted some content around under `data/data/` to get it all into the same tree.  And I've pushed the Terraform content, which is not part of the asset graph, down into `data/data/terraform`.  Still some work to go, but I'm pretty happy with how things are shaping up.  For example, the granular template dependencies, as shown [here][1], make it really easy to audit for "are we providing more template parameters than the template actually needs?" and avoids template rebuilds when assets that are unrelated to a given template are updated.  The much-expanded graph (now also oriented left-to-right) is [here][2] (although I'm still only at 26 of the ~62 paths from the master `bootstrap.ign`).

[1]: https://github.com/openshift/installer/blob/0fb0b465a13b1c78ef7aff15c89657319fd0aa4f/pkg/installerassets/manifests.go
[2]: https://github.com/openshift/installer/blob/0fb0b465a13b1c78ef7aff15c89657319fd0aa4f/docs/user/assets.svg

**Comment by wking** [comment]
I've pushed 0fb0b46 -> 2aa5253, rebasing onto the current master and filling in a bunch of machine-config stuff for the new approach (2aa5253 itself is just a workaround until #730 lands).  Now you can launch a cluster on libvirt :).  Or at least, enough to make Terraform happy.  The cluster doesn't come up, but with all the CI flakes we've had over the last two days I didn't dig too deeply into why.  Once CI gets healthy again, I'll give this another pass to see how healthy the libvirt cluster actually is.

And while I still have some AWS and OpenStack stuff to fill in:

```console
$ git diff --shortstat origin/master..HEAD -- cmd data pkg | cat
333 files changed, 8804 insertions(+), 12901 deletions(-)
```

so I must be doing something right ;).

**Comment by wking** [comment]
I still like this approach, but I'm giving up on rebasing it ;).

/close

**Pull request closed without merging**